extra_data/components.py

```diff
@@ -425,6 +426,118 @@ def write_virtual_cxi(self, filename):
         from .write_cxi import VirtualCXIWriter
         VirtualCXIWriter(self).write(filename)
 
+    def write_frames(self, filename, trains, pulses):
+        """Write selected detector frames to a new EuXFEL HDF5 file
+
+        trains and pulses should be 1D arrays of the same length, containing
+        train IDs and pulse IDs (corresponding to the pulse IDs recorded by
+        the detector). i.e. (trains[i], pulses[i]) identifies one frame.
+        """
+        if (trains.ndim != 1) or (pulses.ndim != 1):
+            raise ValueError("trains & pulses must be 1D arrays")
+        inc_tp_ids = zip_trains_pulses(trains, pulses)
+
+        writer = FramesFileWriter(filename, self.data, inc_tp_ids)
+        try:
+            writer.write()
+        finally:
+            writer.file.close()
+
+
+def zip_trains_pulses(trains, pulses):
+    """Combine two similar arrays of train & pulse IDs as one struct array
+    """
+    if trains.shape != pulses.shape:
+        raise ValueError(
+            f"Train & pulse arrays don't match ({trains.shape} != {pulses.shape})"
+        )
+
+    res = np.zeros(trains.shape, dtype=np.dtype([
+        ('trainId', np.uint64), ('pulseId', np.uint64)
+    ]))
+    res['trainId'] = trains
+    res['pulseId'] = pulses
+    return res
+
+
+class FramesFileWriter(FileWriter):
+    """Write selected detector frames in European XFEL HDF5 format"""
+    def __init__(self, path, data, inc_tp_ids):
+        super().__init__(path, data)
+        self.inc_tp_ids = inc_tp_ids
+
+    def prepare_source(self, source):
+        """Prepare all the datasets for one source.
+
+        We do this as a separate step so the contents of the file are defined
+        together before the main data.
+        """
+        if source not in self.data.instrument_sources:
+            return super().prepare_source(source)
+
+        for key in sorted(self.data.keys_for_source(source)):
+            path = 'INSTRUMENT/{}/{}'.format(source, key.replace('.', '/'))
+            if key.startswith('image.'):
+                nentries = 0  # dataset grows later
+            else:
+                nentries = self.data.get_data_counts(source, key).sum()
```

Comment:
It seem you need override only this part in entire method. Why do not put this in class member, like `_get_number_of_data_entries(source, key)`?

Reply:
I could do that, but I think it makes an uncomfortably tight coupling between the subclass and its parent. Returning 0 from `_get_number_of_data_entries()` would be a convenient lie - it's not the real number of entries, it's just useful to pretend it is when we set up the dataset.

So using that method in any other context would go wrong, but only when you're using the subclass. If you're just looking at the parent class, it would seem entirely reasonable to use `_get_number_of_data_entries()` in different contexts.

So, IMO, duplicating a few lines in this method is preferable to that. Another option would be to find the real number of entries for image data, but that means extra complexity as you handle all of the pulseId data again.

Reply:
Yes, the name of method may confuse. Call it like _guess_number_of_storing_entries() instead or any similar way. It will precisely define a possible context.

I don't want to say that any one line could never be repeated. But the good portion of logic is repeated here. Any future developers have to remember that they must change it in two places.

I also don't see too tight coupling. Suggested interface will be introduced in base class FileWriter as:
```python
def _guess_number_of_entries_to_record(source, key):
    return self.data.get_data_counts(source, key).sum()
```
Usually you don't need to redefine anything in children. When it necessary, you have an option to change `_guess_number_of_entries_to_record` at any children like here in FramesFileWriter:
```python
def _guess_number_of_entries_to_record(this, source, key):
    if key.startswith('image.'):
        return 0
    else:
        return super()._guess_number_of_entries_to_record(source, key)
```
keeping prepare_source() unchanged.

Or get rid of this hook at all by redefining prepare_source().

Looks very flexible.

Actually, IMO, the ideal solution is different. It would be good having an option to create a collection which exposes part of the real data filtered by the list of frames. Then no special recorder is needed. But implementation may be too hard at the moment.

So, I think this is a little, cheap and good improvement. Since it works as is you may merge it without changes if I haven't managed to convince you.

Reply:
Fair enough, I'm persuaded. I've attributed you on the commit, as I copied it from the code in your commen.t